framework/src/org/checkerframework/framework/type/AnnotatedTypeMirror.java

```diff
@@ -925,29 +911,34 @@ public void setTypeArguments(List<? extends AnnotatedTypeMirror> ts) {
 
         /** @return the type argument for this type */
         public List<AnnotatedTypeMirror> getTypeArguments() {
-            if (typeArgs == null) {
-                typeArgs = new ArrayList<AnnotatedTypeMirror>();
-                if (wasRaw()) {
-                    // TODO: This doesn't handle recursive type parameter
-                    // e.g. class Pair<Y extends List<Y>> { ... }
-                    // Type argument inference for raw types can be improved. See Issue #635.
-                    // https://github.com/typetools/checker-framework/issues/635
-                    AnnotatedDeclaredType declaration =
-                            atypeFactory.fromElement((TypeElement) getUnderlyingType().asElement());
-                    for (AnnotatedTypeMirror typeParam : declaration.getTypeArguments()) {
-                        AnnotatedWildcardType wct =
-                                atypeFactory.getUninferredWildcardType(
-                                        (AnnotatedTypeVariable) typeParam);
-                        typeArgs.add(wct);
-                    }
-                } else if (!getUnderlyingType().getTypeArguments().isEmpty()) { // lazy init
-                    for (TypeMirror t : getUnderlyingType().getTypeArguments()) {
-                        typeArgs.add(createType(t, atypeFactory, declaration));
-                    }
+            if (typeArgs != null) {
+                return typeArgs;
+            } else if (wasRaw()) {
+                // Initialize the type arguments with uninferred wildcards
+                BoundsInitializer.initializeTypeArgs(this);
+
+                // Copy annotations from the declaration to the wildcards.
```

Comment:
Could this be moved into `initializeTypeArgs`? It seems strange to first have a dedicated function to initialize type arguments and then modify the result again. If not, please add a comment why this logic needs to be here and not there.

Reply:
The BoundsInitializer creates types without adding annotations, so this code doesn't belong there.